File: Data Structures & Algorithms/eating-bananas/submission-1.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int n = piles.size();
        int sum = 0;
        int maxK = 0;

        for (int i = 0; i < n; i++) {
            maxK = max(maxK, piles[i]);
            sum += piles[i];
        }

        int minK = 1;
        int curBest = maxK;
        int curBestHours = sum;
        while (minK <= maxK) {
            int mid = minK + (maxK - minK) / 2;
            int hours = getHours(piles, mid);

            cout << "Min: " << minK << ", Max: " << maxK << ", Mid: " << mid << ", Hours: " << hours << ", CurBest: " << curBest << "\n";

            if (hours <= h) {
                if (mid < curBest) {
                    curBestHours = hours;
                    curBest = mid;
                    cout << "New best: " << mid << "\n";
                }
                maxK = mid - 1;
            } else {
                minK = mid + 1;
            }

            cout << "\n";
        }

        return curBest;
    }

    int getHours(vector<int>& piles, int k) {
        int hours = 0;
        for (int i = 0; i < piles.size(); i++) {
            hours += piles[i] / k;
            hours += piles[i] % k > 0 ? 1 : 0;
        }
        return hours;
    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l = 0
        r = len(nums) - 1

        if len(nums) == 1:
            return nums[0]

        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid - 1] > nums[mid]:
                return nums[mid]
            if nums[l - 1] > nums[l]:
                return nums[l]
            
            if nums[mid] < nums[l]:
                r = mid - 1
            else:
                l = mid + 1